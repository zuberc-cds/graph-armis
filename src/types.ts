// Providers often supply types with their API libraries.

export interface AcmeUser {
  id: string;
  name: string;
}

export interface AcmeGroup {
  id: string;
  name: string;
  users?: Pick<AcmeUser, 'id'>[];
}

export interface ArmisDevice {
  id: number;
  name: string;
  category: string;
  type: string;
  model: string;
  manufacturer: string;
  operatingSystem: string;
  operatingSystemVersion: string;
  lastSeen: string;
}

export interface ArmisSite {
  id: number;
  lat: number;
  lng: number;
  location: string;
  name: string;
  parentId: string;
}

export interface ArmisAlert extends ArmisFinding {
  alertId: number;
  title: string;
  classification: string;
  type: string;
  policyId: number;
  policyTitle: string;
  time: string;
  deviceIds: number[];
}

export interface ArmisVulnerability extends ArmisFinding {
  cveUid: string;
  affectedDevicesCount: number;
  attackComplexity: string;
  attackVector: string;
  availabilityImpact: string;
  cvssScore: number;
  epssPercentile: number;
  epssScore: number;
  exploitabilityScore: number;
  impactScore: number;
  integrityImpact: string;
  publishedDate: string;
  score: number;
  commonName: string;
  threatTags: string[];
}

export interface ArmisFinding {
  id: string;
  description: string;
  severity: string;
  status: string;
}

export type ArmisDeviceVulnerability = {
  cveUid: string;
};

// Those can be useful to a degree, but often they're just full of optional
// values. Understanding the response data may be more reliably accomplished by
// reviewing the API response recordings produced by testing the wrapper client
// (./client.ts). However, when there are no types provided, it is necessary to define
// opaque types for each resource, to communicate the records that are expected
// to come from an endpoint and are provided to iterating functions.

/*
import { Opaque } from 'type-fest';
export type AcmeUser = Opaque<any, 'AcmeUser'>;
export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
*/
