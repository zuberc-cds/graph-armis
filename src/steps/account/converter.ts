import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';

export function createAccountEntity(): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {
        id: 'armis-unique-account-id',
        name: 'Example Co. Armis Account',
      },
      assign: {
        _key: 'armis-unique-account-id',
        _type: Entities.ACCOUNT._type,
        _class: Entities.ACCOUNT._class,
        mfaEnabled: true,
        // This is a custom property that is not a part of the data model class
        // hierarchy. See: https://github.com/JupiterOne/data-model/blob/master/src/schemas/Account.json
        manager: 'Manager Name',
      },
    },
  });
}
