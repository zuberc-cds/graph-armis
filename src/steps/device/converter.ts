import {
  createIntegrationEntity,
  createDirectRelationship,
  Entity,
  RelationshipClass,
  Relationship,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';
import { AcmeGroup, AcmeUser, ArmisDevice } from '../../types';

export function createDeviceEntity(sourceDevice: ArmisDevice): Entity {
  /* eslint-disable no-console */
  return createIntegrationEntity({
    entityData: {
      source: sourceDevice,
      assign: {
        _type: Entities.DEVICE._type,
        _class: Entities.DEVICE._class,
        _key: 'armis-device-' + sourceDevice.id.toString(),
        id: 'armis-device-' + sourceDevice.id.toString(),
        deviceId: sourceDevice.id.toString(),
        category: 'asset',
        model: sourceDevice.model !== null ? sourceDevice.model : '',
        make: sourceDevice.manufacturer,
        lastSeen: sourceDevice.lastSeen,
        serial: sourceDevice.name,
        name:
          sourceDevice.name !== null
            ? sourceDevice.name
            : sourceDevice.id.toString(),
        displayName: sourceDevice.name,
        osName: sourceDevice.operatingSystem
          ? sourceDevice.operatingSystem
          : '',
        osVersion: sourceDevice.operatingSystemVersion
          ? sourceDevice.operatingSystemVersion
          : '',
      },
    },
  });
}

export function createUserEntity(user: AcmeUser): Entity {
  return createIntegrationEntity({
    entityData: {
      source: user,
      assign: {
        _type: Entities.USER._type,
        _class: Entities.USER._class,
        _key: user.id,
        username: 'testusername',
        email: 'test@test.com',
        active: true, // this is a required property
        // This is a custom property that is not a part of the data model class
        // hierarchy. See: https://github.com/JupiterOne/data-model/blob/master/src/schemas/User.json
        firstName: 'John',
      },
    },
  });
}

export function createGroupEntity(group: AcmeGroup): Entity {
  return createIntegrationEntity({
    entityData: {
      source: group,
      assign: {
        _type: Entities.GROUP._type,
        _class: Entities.GROUP._class,
        _key: group.id,
        email: 'testgroup@test.com',
        // This is a custom property that is not a part of the data model class
        // hierarchy. See: https://github.com/JupiterOne/data-model/blob/master/src/schemas/UserGroup.json
        logoLink: 'https://test.com/logo.png',
      },
    },
  });
}

export function createAccountUserRelationship(
  account: Entity,
  user: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: user,
  });
}
export function createAccountGroupRelationship(
  account: Entity,
  group: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: group,
  });
}

export function createGroupUserRelationship(
  group: Entity,
  user: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: group,
    to: user,
  });
}

export function createAccountDeviceRelationship(
  account: Entity,
  device: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.MANAGES,
    from: account,
    to: device,
  });
}
