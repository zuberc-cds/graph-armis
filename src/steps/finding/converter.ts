import {
  createIntegrationEntity,
  createDirectRelationship,
  Entity,
  RelationshipClass,
  Relationship,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';
import {
  AcmeGroup,
  AcmeUser,
  ArmisAlert,
  ArmisFinding,
  ArmisVulnerability,
} from '../../types';

/**
 * Creates an Entity object for an Alert finding.
 *
 * @param sourceAlert - The source alert from which to create the Entity.
 * @returns The created Entity object.
 */
export function createAlertFindingEntity(sourceAlert: ArmisAlert): Entity {
  // Map of severity levels
  const severityMap = ['Low', 'Medium', 'High', 'Critical'];

  return createIntegrationEntity({
    entityData: {
      // Set the common properties for the Entity
      source: sourceAlert,
      assign: {
        _type: Entities.FINDING_ALERT._type,
        _class: Entities.FINDING_ALERT._class,
        _key: 'armis-alert-' + sourceAlert.alertId,
        category: sourceAlert.type,
        severity: sourceAlert.severity,
        numericSeverity:
          // Convert severity to numeric value (0 for 'Unavailable' or null)
          sourceAlert.severity === 'Unavailable' ||
          sourceAlert.severity === null
            ? 0
            : severityMap.indexOf(sourceAlert.severity),
        open: false,
        description:
          // Set description to empty string if null
          sourceAlert.description != null ? sourceAlert.description : '',
        status: sourceAlert.status,
        reportedOn: parseTimePropertyValue(sourceAlert.time),
        name: sourceAlert.title,
        policyTitle: sourceAlert.policyTitle,
        policyId: sourceAlert.policyId,
      },
    },
  });
}

/**
 * Creates a vulnerability finding entity based on the provided source vulnerability.
 * @param sourceVulnerability - The source vulnerability object.
 * @returns The created entity.
 */
export function createVulnerabilityFindingEntity(
  sourceVulnerability: ArmisVulnerability,
): Entity {
  // Map of severity values
  const severityMap = ['Low', 'Medium', 'High', 'Critical'];

  return createIntegrationEntity({
    entityData: {
      source: sourceVulnerability,
      assign: {
        _type: Entities.FINDING_VULNERABILITY._type,
        _class: Entities.FINDING_VULNERABILITY._class,
        _key: 'armis-vulnerability-' + sourceVulnerability.id,
        category: sourceVulnerability.attackVector,
        severity: sourceVulnerability.severity,
        blocking:
          sourceVulnerability.impactScore >= 9 &&
          sourceVulnerability.severity === 'Critical'
            ? true
            : false,
        numericSeverity:
          sourceVulnerability.severity === 'Unavailable' ||
          sourceVulnerability.severity === null
            ? 0
            : severityMap.indexOf(sourceVulnerability.severity),
        open: sourceVulnerability.status === 'Open' ? true : false,
        description:
          sourceVulnerability.description != null
            ? sourceVulnerability.description
            : '',
        status: sourceVulnerability.status,
        reportedOn: parseTimePropertyValue(sourceVulnerability.publishedDate),
        name: 'armis-vulnerability-' + sourceVulnerability.id,
        displayName: 'armis-vulnerability-' + sourceVulnerability.id,
        production: true,
        public:
          sourceVulnerability.threatTags != null
            ? sourceVulnerability.threatTags.indexOf('Publicly Available') !==
              -1
              ? true
              : false
            : false,
        score: sourceVulnerability.score,
        impact: sourceVulnerability.impactScore,
        exploitability: sourceVulnerability.exploitabilityScore,
        vector: sourceVulnerability.attackVector,
        impacts: ['devices'],
      },
    },
  });
}

/**
 * Creates a new Finding entity based on the provided source finding.
 * @param sourceFinding - The source finding object.
 * @returns The newly created Entity.
 */
export function createFindingEntity(sourceFinding: ArmisFinding): Entity {
  const severityMap = ['Low', 'Medium', 'High', 'Critical'];

  return createIntegrationEntity({
    entityData: {
      source: sourceFinding,
      assign: {
        _type: Entities.FINDING._type,
        _class: Entities.FINDING_ALERT._class,
        _key: 'armis-finding-' + sourceFinding.id,
        name: 'armis-finding-' + sourceFinding.id,
        displayName: 'armis-finding-' + sourceFinding.id,
        category: 'armis-finding',
        severity: sourceFinding.severity,
        numericSeverity:
          sourceFinding.severity === 'Unavailable' ||
          sourceFinding.severity === null
            ? 0
            : severityMap.indexOf(sourceFinding.severity),

        description:
          sourceFinding.description != null ? sourceFinding.description : '',
        status: sourceFinding.status,
        open: sourceFinding.status === 'Open' ? true : false,
      },
    },
  });
}

/**
 * Creates a user entity based on the provided user object.
 *
 * @param {AcmeUser} user - The user object used as the source for creating the entity.
 * @return {Entity} The created user entity.
 */
export function createUserEntity(user: AcmeUser): Entity {
  return createIntegrationEntity({
    entityData: {
      source: user,
      assign: {
        _type: Entities.USER._type,
        _class: Entities.USER._class,
        _key: user.id,
        username: 'testusername',
        email: 'test@test.com',
        active: true, // this is a required property
        // This is a custom property that is not a part of the data model class
        // hierarchy. See: https://github.com/JupiterOne/data-model/blob/master/src/schemas/User.json
        firstName: 'John',
      },
    },
  });
}

/**
 * Creates a group entity.
 *
 * @param {AcmeGroup} group - The group object used to create the entity.
 * @return {Entity} The created entity.
 */
export function createGroupEntity(group: AcmeGroup): Entity {
  return createIntegrationEntity({
    entityData: {
      source: group,
      assign: {
        _type: Entities.GROUP._type,
        _class: Entities.GROUP._class,
        _key: group.id,
        email: 'testgroup@test.com',
        // This is a custom property that is not a part of the data model class
        // hierarchy. See: https://github.com/JupiterOne/data-model/blob/master/src/schemas/UserGroup.json
        logoLink: 'https://test.com/logo.png',
      },
    },
  });
}

/**
 * Creates a relationship between a device and a finding.
 *
 * @param {Entity} device - The device to create the relationship for.
 * @param {Entity} finding - The finding to create the relationship with.
 * @return {Relationship} The created relationship between the device and the finding.
 */
export function createDeviceFindingRelationship(
  device: Entity,
  finding: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: device,
    to: finding,
  });
}

/**
 * Creates a relationship between a finding and a vulnerability.
 *
 * @param {Entity} finding - The finding entity.
 * @param {Entity} vulnerability - The vulnerability entity.
 * @return {Relationship} The created relationship.
 */
export function createFindingVulnerabilityRelationship(
  finding: Entity,
  vulnerability: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.IS,
    from: finding,
    to: vulnerability,
  });
}

/**
 * Creates a relationship between an account and a user.
 *
 * @param {Entity} account - The account entity.
 * @param {Entity} user - The user entity.
 * @return {Relationship} The created relationship.
 */
export function createAccountUserRelationship(
  account: Entity,
  user: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: user,
  });
}

/**
 * Creates a relationship between an account and a group.
 *
 * @param {Entity} account - The account entity.
 * @param {Entity} group - The group entity.
 * @return {Relationship} The created relationship between the account and the group.
 */
export function createAccountGroupRelationship(
  account: Entity,
  group: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: group,
  });
}

/**
 * Creates a group-user relationship.
 *
 * @param {Entity} group - The group entity.
 * @param {Entity} user - The user entity.
 * @return {Relationship} The created relationship.
 */
export function createGroupUserRelationship(
  group: Entity,
  user: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: group,
    to: user,
  });
}
