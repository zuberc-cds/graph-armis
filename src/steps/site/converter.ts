import {
  createIntegrationEntity,
  createDirectRelationship,
  Entity,
  RelationshipClass,
  Relationship,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';
import { ArmisSite } from '../../types';

/**
 * Creates a site entity based on the provided source site.
 *
 * @param {ArmisSite} sourceSite - The source site to create the entity from.
 * @return {Entity} The created site entity.
 */
export function createSiteEntity(sourceSite: ArmisSite): Entity {
  return createIntegrationEntity({
    entityData: {
      source: sourceSite,
      assign: {
        _type: Entities.SITE._type,
        _class: Entities.SITE._class,
        _key: sourceSite.name + '-' + sourceSite.id.toString(),
        id: sourceSite.id.toString(),
        location: sourceSite.location,
      },
    },
  });
}

/**
 * Creates a relationship between an account and a user.
 *
 * @param {Entity} account - The account entity.
 * @param {Entity} user - The user entity.
 * @return {Relationship} The created relationship.
 */
export function createAccountUserRelationship(
  account: Entity,
  user: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: user,
  });
}

/**
 * Create an account-group relationship.
 *
 * @param {Entity} account - The account entity.
 * @param {Entity} group - The group entity.
 * @return {Relationship} The created relationship.
 */
export function createAccountGroupRelationship(
  account: Entity,
  group: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: group,
  });
}

/**
 * Creates a group-user relationship.
 *
 * @param {Entity} group - The group entity.
 * @param {Entity} user - The user entity.
 * @return {Relationship} The created relationship.
 */
export function createGroupUserRelationship(
  group: Entity,
  user: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: group,
    to: user,
  });
}

/**
 * Creates a relationship between an account and a site.
 *
 * @param {Entity} account - The account entity.
 * @param {Entity} site - The site entity.
 * @return {Relationship} The created relationship.
 */
export function createAccountSiteRelationship(
  account: Entity,
  site: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: site,
  });
}
