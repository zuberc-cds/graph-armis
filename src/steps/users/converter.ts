import {
  createIntegrationEntity,
  createDirectRelationship,
  Entity,
  RelationshipClass,
  Relationship,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';
import { ArmisUser } from '../../types';

export function createUserEntity(sourceUser: ArmisUser): Entity {
  return createIntegrationEntity({
    entityData: {
      source: sourceUser,
      assign: {
        _type: Entities.USER._type,
        _class: Entities.USER._class,
        _key: 'armis-user-' + sourceUser.id.toString(),
        id: sourceUser.id.toString(),
        email: sourceUser.email,
        mfaEnabled: sourceUser.twoFactorAuthentication,
        active: sourceUser.isActive,
        username: sourceUser.username,
        location: sourceUser.location,
      },
    },
  });
}

export function createAccountUserRelationship(
  account: Entity,
  user: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: user,
  });
}
