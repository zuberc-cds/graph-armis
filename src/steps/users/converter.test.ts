import { ArmisUser } from '../../types';
import { createUserEntity } from './converter';

describe('#createUserEntity', () => {
  test('should convert to user entity', () => {
    const userEntity: ArmisUser = {
      id: 1,
      email: 'test@gmail.com',
      isActive: true,
      lastLoginTime: '22/02/2020',
      location: 'USA',
      name: 'Test',
      phone: 9876543210,
      reportPermissions: '',
      role: 'admin',
      roleAssignment: [
        {
          name: ['read'],
          sites: ['crest'],
        },
      ],
      title: 'test',
      twoFactorAuthentication: true,
      username: 'dummy-user',
    };
    expect(createUserEntity(userEntity)).toMatchSnapshot();
  });
});
