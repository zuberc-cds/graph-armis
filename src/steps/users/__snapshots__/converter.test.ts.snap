// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createUserEntity should convert to user entity 1`] = `
Object {
  "_class": Array [
    "User",
  ],
  "_key": "armis-user-1",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "email": "test@gmail.com",
        "id": 1,
        "isActive": true,
        "lastLoginTime": "22/02/2020",
        "location": "USA",
        "name": "Test",
        "phone": 9876543210,
        "reportPermissions": "",
        "role": "admin",
        "roleAssignment": Array [
          Object {
            "name": Array [
              "read",
            ],
            "sites": Array [
              "crest",
            ],
          },
        ],
        "title": "test",
        "twoFactorAuthentication": true,
        "username": "dummy-user",
      },
    },
  ],
  "_type": "armis_user",
  "active": true,
  "createdOn": undefined,
  "displayName": "Test",
  "email": "test@gmail.com",
  "id": "1",
  "location": "USA",
  "mfaEnabled": true,
  "name": "Test",
  "username": "dummy-user",
}
`;
